### PR TITLE
tbcd: clean up peer and peermanager

### LINKED ISSUE #214: Header dowloads can stall
### Confirmation

- [X] I have checked for similar issues.

### Describe the bug

During IBC blockheader download phase sometimes the code stalls.

### Expected behaviour

Recover from a bad peer if the header download stalls.

### Environment

Version: 
Operating System: 
Architecture: 

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by marcopeereboom**
**Summary**
Peer manager is currently a turd because it was part of the server struct and got adulterated over time.

**Changes**
 Clean up this mess by making it a full citizen.

Fixes #214

### CODE REVIEW

File: service/tbc/peer_manager.go [hunk 1]
```diff
@@ -5,129 +5,349 @@
...
+	defer log.Tracef("Random exit")
 
-	log.Debugf("PeerDelete exit good %v bad %v", allGoodPeers, allBadPeers)
+	pm.mtx.RLock()
+	pm.mtx.RUnlock()
+	for _, p := range pm.peers {
+		if p.isConnected() {
+			return p, nil
+		}
+	}
```
Comment (joshuasing) [inline service/tbc/peer_manager.go:260]: `pm.mtx.RUnlock()` is called before reading `pm.peers`, which results in a race condition.
Did you mean to copy it before unlocking, then iterate over the copied map?
```diff
import (
+	"maps"
)
```

```suggestion
	pm.mtx.RLock()
	peers := maps.Clone(pm.peers)
	pm.mtx.RUnlock()
	for _, p := range peers {
		if p.isConnected() {
			return p, nil
		}
	}
```

File: service/tbc/peer_manager.go [hunk 2]
```diff
@@ -5,129 +5,349 @@
...
-	}
+	pm.mtx.Unlock()
 
-	allGoodPeers := len(pm.peersGood)
-	allBadPeers := len(pm.peersBad)
+	return nil
+}
 
-	pm.peersMtx.Unlock()
+func (pm *PeerManager) Random() (*peer, error) {
```
Comment (joshuasing) [inline service/tbc/peer_manager.go:250]:
```suggestion
// Random returns a random connected peer.
func (pm *PeerManager) Random() (*peer, error) {
```

File: service/tbc/peer_manager.go [hunk 3]
```diff
@@ -5,129 +5,349 @@
...
+	pm.mtx.RLock()
+	pm.mtx.RUnlock()
+	for _, p := range pm.peers {
+		if p.isConnected() {
+			return p, nil
+		}
+	}
 
-	return nil
+	return nil, errors.New("no peers")
```
Comment (joshuasing) [inline service/tbc/peer_manager.go:262]:
```suggestion
	return nil, errors.New("no connected peers")
```

**Review by joshuasing — CHANGES_REQUESTED** [review]

File: service/tbc/peer_manager.go [hunk 4]
```diff
@@ -5,129 +5,349 @@
...
+	peers map[string]*peer // connected peers
+	good  map[string]struct{}
+	bad   map[string]struct{}
 }
 
-// newPeerManager returns a new peer manager.
-func newPeerManager() *PeerManager {
+// NewPeerManager returns a new peer manager.
+func NewPeerManager(net wire.BitcoinNet, want int) (*PeerManager, error) {
+	if want == 0 {
```
Comment (ClaytonNorthey92) [inline service/tbc/peer_manager.go:59]: should we check that this is `> 0` ? or make it a `uint`?

File: service/tbc/peer_manager.go [hunk 5]
```diff
@@ -5,129 +5,349 @@
...
+		return nil, errors.New("peers wanted must not be 0")
+	}
+
+	var dnsSeeds []string
+	switch net {
```
Comment (ClaytonNorthey92) [inline service/tbc/peer_manager.go:64]: what about `Regtest` / `localnet`? I believe we need to add a case for those seeds.

File: service/tbc/peer_manager.go [hunk 6]
```diff
@@ -5,129 +5,349 @@
...
-	allBadPeers := len(pm.peersBad)
-	pm.peersMtx.Unlock()
 
-	log.Debugf("PeersInsert exit %v good %v bad %v",
-		len(peers), allGoodPeers, allBadPeers)
+	pm.mtx.Lock()
+	defer pm.mtx.Unlock()
+
+	// If peer is connected don't add it to good list
+	if _, ok := pm.peers[address]; ok {
```
Comment (ClaytonNorthey92) [inline service/tbc/peer_manager.go:182]: is this an _error_ or should it just be a no-op?

File: service/tbc/peer.go [hunk 7]
```diff
@@ -211,10 +211,16 @@ func (p *peer) close() error {
 	log.Tracef("close")
 	defer log.Tracef("close exit")
 
+	if p == nil {
+		panic("p")
+	}
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply service/tbc/peer.go:229]: It was debugging and is now removed.

File: service/tbc/peer_manager.go (hunk 3 above)
Comment (marcopeereboom) [reply service/tbc/peer_manager.go:260]: Fixed.

File: service/tbc/peer_manager.go (hunk 4 above)
Comment (marcopeereboom) [reply service/tbc/peer_manager.go:62]: Fixed

File: service/tbc/peer_manager.go (hunk 6 above)
Comment (marcopeereboom) [reply service/tbc/peer_manager.go:187]: Fixed.

File: service/tbc/peer_manager.go (hunk 5 above)
Comment (marcopeereboom) [reply service/tbc/peer_manager.go:67]: No. We should use: `TBC_SEEDS="127.0.0.1:18333"` to make localnet work. Let me know if this isn't good enough.

File: service/tbc/peermanager_test.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+			return nil
+		}
+	}
+}
+
+func TestPeerManager(t *testing.T) {
+	t.Skip("this test connects to testnet3")
+	want := 2
+	wantLoop := want * 2
+	pm, err := NewPeerManager(wire.TestNet3, []string{}, want)
```
Comment (ClaytonNorthey92) [inline service/tbc/peermanager_test.go:41]: @marcopeereboom does this connect to testnet3 during the test? it seemed like it did when I ran it so I skipped this as I don't want our tests to depend on testnet3 itself.

Comment (marcopeereboom) [reply service/tbc/peermanager_test.go:41]: yeah i think that's the bug. Need to provide 127.0.0.1 in seeds

File: service/tbc/peer_manager.go (hunk 6 above)
Comment (marcopeereboom) [reply service/tbc/peer_manager.go:187]: error for diagnostic.

File: service/tbc/peer_manager.go [hunk 9]
```diff
@@ -5,129 +5,411 @@
...
 
-	allGoodPeers := len(pm.peersGood)
-	allBadPeers := len(pm.peersBad)
+func (pm *PeerManager) connect(ctx context.Context, p *peer) error {
+	log.Tracef("connect: %v %v", p.Id(), p)
+	defer log.Tracef("connect exit: %v %v", p.Id(), p)
 
-	pm.peersMtx.Unlock()
+	err := p.connect(ctx)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply service/tbc/peer_manager.go:334]: No, we use the `err` later.

[the remark this answers is not in the document]
Comment (joshuasing) [reply service/tbc/peer_manager.go]: Oops, you are right. Sorry

**Review by joshuasing — APPROVED** [review]

**Review by ClaytonNorthey92 — APPROVED** [review]

**Pull request merged**